src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1057 @@
+        struct cpu_msg msg = {VPLIC_IPI_ID, UPDATE_HART_LINE};
+        cpu_send_msg(pcpu_id, &msg);       
+    }
+}
+
+/**
+ * @brief Processes an incoming event.
+ * 
+ * @param event the event id
+ * @param data
+ */
+static void vaplic_ipi_handler(uint32_t event, uint64_t data) 
+{
+    switch(event) {
+        case UPDATE_HART_LINE:
+            vaplic_update_hart_line(cpu()->vcpu);
+            break;
+    }
+}
+
+/** APLIC Functions emulation */
+
+/**
+ * @brief Write to domaincfg register a new value.
+ * 
+ * @param vcpu 
+ * @param new_val The new value to write to domaincfg
+ */
+static void vaplic_set_domaincfg(struct vcpu *vcpu, uint32_t new_val){
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    /** Update only the virtual domaincfg */
+    /** Only Interrupt Enable is configurable. */
+    new_val &= APLIC_DOMAINCFG_IE;
+    virqc->domaincfg = new_val | APLIC_DOMAINCFG_RO80;
+    spin_unlock(&virqc->lock);
+
+    vaplic_update_hart_line(vcpu);
+}
+
+/**
+ * @brief Read from domaincfg
+ * 
+ * @param vcpu 
+ * @return uint32_t domaincfg value 
+ */
+static uint32_t vaplic_get_domaincfg(struct vcpu *vcpu){
+    uint32_t ret = 0;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    ret = virqc->domaincfg;
+    return ret;
+}
+
+/**
+ * @brief Read from a given interrupt sourcecfg register
+ * 
+ * @param vcpu 
+ * @param intp_id Interrupt id to read
+ * @return uint32_t 
+ */
+static uint32_t vaplic_get_sourcecfg(struct vcpu *vcpu, irqid_t intp_id){
+    uint32_t real_int_id = intp_id - 1;
+    uint32_t ret = 0;
+
+    if(intp_id == 0)
+        return ret;
+    struct virqc * virqc = &vcpu->vm->arch.virqc;
+    if (real_int_id < APLIC_MAX_INTERRUPTS) ret = virqc->srccfg[real_int_id];
+    return ret;
+}
+
+/**
+ * @brief Write to sourcecfg register of a given interrupt
+ * 
+ * @param vcpu 
+ * @param intp_id interrupt id to write to
+ * @param new_val value to write to sourcecfg
+ */
+static void vaplic_set_sourcecfg(struct vcpu *vcpu, irqid_t intp_id, uint32_t new_val){
+    struct virqc *virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    /** If intp is valid and new source config is different from prev. one.*/
+    if (intp_id > 0 && intp_id < APLIC_MAX_INTERRUPTS && 
+        vaplic_get_sourcecfg(vcpu, intp_id) != new_val) {
+        /** If intp is being delegated make whole reg 0.
+         *  This happens because this domain is always a leaf. */        
+        new_val &= (new_val & APLIC_SRCCFG_D) ? 0 : APLIC_SRCCFG_SM;
+
+        /** If SM is reserved make intp inactive */
+        if(new_val == 2 || new_val == 3)
+            new_val = APLIC_SOURCECFG_SM_INACTIVE;
+        
+        /** Only edge sense can be virtualized for know */
+        if(new_val  == APLIC_SOURCECFG_SM_LEVEL_HIGH){
+            new_val = APLIC_SOURCECFG_SM_EDGE_RISE;
+        } else if (new_val  == APLIC_SOURCECFG_SM_LEVEL_LOW){
+            new_val = APLIC_SOURCECFG_SM_EDGE_FALL;
+        }
```

Comment:
I translate level interrupts to edge because Linux configures some interrupts (UART) in level high mode (even though in the dtb I specify that the UART is rising edged). So, if we translate it to inactive, Linux will never print anything. 

Now that I am thinking, I need to look at the latest aia version of Linux and check if the same is happening, otherwise, I could probably send a PR to fix it. Just to be sure, in section _4.5.2 Source configurations (sourcecfg[1]–sourcecfg[1023])_ from AIA spec, you interpret it the same way as I do, right? Microarchitecture can choose which SMs it supports (aside from inactive which is mandatory).

Reply:
> you interpret it the same way as I do, right? Microarchitecture can choose which SMs it supports (aside from inactive which is mandatory).

Yes but I think this is also related to how the device signals the interrupt. Being high-level might have semantics that cannot be replicated with edge triggered.